rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,268 @@
+  - add `ResourceDownloadError(root_cid: CID, error: String)` :: Informs the daemon that a requested resource could not be downloaded.
+
+### Migrating Mina Blocks to Bitswap
+
+To migrate Mina block propagation to Bitswap, we will separate a Mina block into 2 portions: a block header, and a block body. Most of the data in a Mina block is stored inside of the `staged_ledger_diff`. The common data in every Mina block is ~8.06kb (including the `protocol_state_proof`), so using everything __except__ for the `staged_ledger_diff` as the block header seems natural. The `staged_ledger_diff` would then act as the block body for Mina blocks, and would be downloaded/made available via Bitswap rather than broadcast over pub/sub.
+
+When blocks are broadcast through the network now, only the block header and a root CID for the `staged_ledger_diff` are in the message. When a node receives a new block header, the node will first verify the `protocol_state_proof` (all public information that needs to be fed in for proof verification will be available in the block header). Once the proof is checked, a node would then download the `staged_ledger_diff` via Bitswap. Once that is downloaded, the node would follow the same pattern right now for generating a breadcrumb by expanding the `staged_ledger` from the parent breadcrumb and the new `staged_ledger_diff`, after which the Mina block will be fully validated. At this point, the breadcrumb is added to the frontier.
+
+One large difference from before, however, is that nodes will rebroadcast the block header to other nodes on the network before the `staged_ledger_diff` is downloaded and verified, in order to avoid increasing block propagation time on the network with the new addition of Bitswap. This change brings some unique problems that we need to solve now, as previously, we wouldn't forward Mina blocks to other nodes until we knew the block was fully valid. In the new world, an adversary could broadcast around the same block header and proof, but swap out the `staged_ledger_diff` root Bitswap block CID with different values to attack the network. In order to prevent this, we must now include a commitment in the snark not only to the target staged ledger hash, but also the root Bitswap block CID of the `staged_ledger_diff` that brings us to that state. This makes the attack more expensive to preform since you need to generate a proof for each `staged_ledger_diff` the adversary wants to broadcast erroneously to the network. In addition to this, we will make a rule such that, if a node ever downloads a `staged_ledger_diff` which does not achieve the target staged ledger hash after application to the parent staged ledger, that node will ban the block producer public key of whoever produced that block. This further removes the incentive to perform this kind of attack, since an adversary doing so would lose their ability to submit blocks to nodes in the future.
```

Comment:
> that node will ban the block producer public key of whoever produced that block.

what if a malicoous middleman swaps out the `staged_ledger_diff`, not the block producer, and it causes the block producer to be banned erroneously? I guess the validating node could also check if the signer of the block signed the fake `staged_ledger_diff`, and if so, then we know the block producer was malicious?  (I'm not super familiar with the mina block format so maybe I'm missing something)

Reply:
> ah does CID have to be a hash supported by IPFS? like sha256? I suppose then we can't use the staged-ledger-hash as is

Exactly

Reply:
> what if a malicoous middleman swaps out the staged_ledger_diff, not the block producer, and it causes the block producer to be banned erroneously?

The idea to prevent this was to require the block producer to include a commitment to the `staged_ledger_diff`'s root Bitswap block CID, so that an intermediate party could not manipulate this value in the block header without making the proof invalid. This also makes this sort of attack expensive to do from the perspective of the block producer since they would need to generate a unique proof for each different `staged_ledger_diff` they want to send out (which is computationally expensive; sort of requires a proof of work per message, in some sense).